File: app/src/main/java/app/bristlecone/LocationEngine.java
package app.bristlecone;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.SharedPreferences;
import android.hardware.GeomagneticField;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.SystemClock;

import org.json.JSONObject;

/**
 * Positioning with an honest fallback chain:
 *   1. GPS (satellite fix) when available and accurate.
 *   2. Network location (Wi-Fi and cell towers) when GPS is out. Flagged as an estimate.
 *   3. Dead reckoning from the last good fix using the step counter, compass and barometer,
 *      with a stride-length model learned on this phone from earlier GPS walks. Flagged as
 *      an estimate with a growing uncertainty circle.
 */
public final class LocationEngine implements LocationListener, SensorEventListener {
    public interface Listener { void onUpdate(JSONObject fix); }

    private static final long GPS_STALE_MS = 12000;
    private static final long NET_STALE_MS = 60000;
    private static final float GOOD_GPS_ACC = 35f;

    private final Context ctx;
    private final Listener listener;
    private final LocationManager lm;
    private final SensorManager sm;
    private final SharedPreferences prefs;
    private final Handler main = new Handler(Looper.getMainLooper());
    private final StrideModel model = new StrideModel();
    private final DeadReckoner dr = new DeadReckoner();

    private boolean running;
    private Location lastGps, lastNet;
    private long lastGpsElapsed, lastNetElapsed;

    // Sensor state
    private final float[] rot = new float[9];
    private final float[] orient = new float[3];
    private double headingMag = Double.NaN;
    private double declination = 0;
    private double baroAlt = Double.NaN, baroAltPrev = Double.NaN;
    private long baroPrevMs;
    private double vSpeed = 0;
    private final long[] stepTimes = new long[16];
    private int stepIdx, stepCount;

    // Training state (steps between two clean GPS fixes)
    private Location trainStart;
    private int trainSteps;
    private double trainAltStart = Double.NaN;

    private final Runnable ticker = new Runnable() {
        @Override public void run() {
            if (!running) return;
            emit();
            main.postDelayed(this, 1000);
        }
    };

    public LocationEngine(Context ctx, Listener listener) {
        this.ctx = ctx.getApplicationContext();
        this.listener = listener;
        this.lm = (LocationManager) ctx.getSystemService(Context.LOCATION_SERVICE);
        this.sm = (SensorManager) ctx.getSystemService(Context.SENSOR_SERVICE);
        this.prefs = ctx.getSharedPreferences("bristlecone_location", Context.MODE_PRIVATE);
        model.deserialize(prefs.getString("stride", null));
    }

    public boolean hasBarometer() { return sm != null && sm.getDefaultSensor(Sensor.TYPE_PRESSURE) != null; }
    public boolean hasStepDetector() { return sm != null && sm.getDefaultSensor(Sensor.TYPE_STEP_DETECTOR) != null; }
    public int strideSamples() { return model.samples(); }
    public double strideAt(double cadence) { return model.predict(cadence, 0); }

    public void resetStrideModel() {
        model.reset();
        prefs.edit().putString("stride", model.serialize()).apply();
    }

    @SuppressLint("MissingPermission")
    public void start() {
        if (running) return;
        running = true;
        // Registered even when a provider is switched off right now: Android starts delivering
        // fixes as soon as the user turns it on, without an app restart.
        java.util.List<String> all = lm.getAllProviders();
        try {
            if (all.contains(LocationManager.GPS_PROVIDER))
                lm.requestLocationUpdates(LocationManager.GPS_PROVIDER, 1000, 0, this, Looper.getMainLooper());
        } catch (Exception ignored) { }
        try {
            if (all.contains(LocationManager.NETWORK_PROVIDER))
                lm.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 5000, 0, this, Looper.getMainLooper());
        } catch (Exception ignored) { }
        try {
            Location g = lm.getLastKnownLocation(LocationManager.GPS_PROVIDER);
            Location n = lm.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
            Location best = g != null ? g : n;
            if (best != null && System.currentTimeMillis() - best.getTime() < 10 * 60 * 1000) {
                dr.anchor(best.getLatitude(), best.getLongitude(),
                        best.getAccuracy() + (System.currentTimeMillis() - best.getTime()) / 1000f,
                        System.currentTimeMillis());
            }
        } catch (Exception ignored) { }
        if (sm != null) {
            Sensor rv = sm.getDefaultSensor(Sensor.TYPE_ROTATION_VECTOR);
            if (rv == null) rv = sm.getDefaultSensor(Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR);
            if (rv != null) sm.registerListener(this, rv, SensorManager.SENSOR_DELAY_UI);
            Sensor st = sm.getDefaultSensor(Sensor.TYPE_STEP_DETECTOR);
            if (st != null) sm.registerListener(this, st, SensorManager.SENSOR_DELAY_FASTEST);
            Sensor pr = sm.getDefaultSensor(Sensor.TYPE_PRESSURE);
            if (pr != null) sm.registerListener(this, pr, SensorManager.SENSOR_DELAY_NORMAL);
        }
        main.post(ticker);
    }

    public void stop() {
        if (!running) return;
        running = false;
        try { lm.removeUpdates(this); } catch (Exception ignored) { }
        if (sm != null) sm.unregisterListener(this);
        main.removeCallbacks(ticker);
        prefs.edit().putString("stride", model.serialize()).apply();
    }

    // ------------------------------------------------------------------ Location callbacks

    @Override public void onLocationChanged(Location loc) {
        long now = SystemClock.elapsedRealtime();
        if (LocationManager.GPS_PROVIDER.equals(loc.getProvider())) {
            lastGps = loc; lastGpsElapsed = now;
            declination = new GeomagneticField((float) loc.getLatitude(), (float) loc.getLongitude(),
                    (float) loc.getAltitude(), System.currentTimeMillis()).getDeclination();
            if (loc.getAccuracy() <= GOOD_GPS_ACC) {
                dr.anchor(loc.getLatitude(), loc.getLongitude(), loc.getAccuracy(), System.currentTimeMillis());
                train(loc);
            }
        } else {
            lastNet = loc; lastNetElapsed = now;
            // Re-anchor dead reckoning to the network fix if it is tighter than our estimate.
            if (!gpsFresh() && (!dr.isAnchored() || loc.getAccuracy() < dr.radius())) {
                dr.anchor(loc.getLatitude(), loc.getLongitude(), loc.getAccuracy(), System.currentTimeMillis());
            }
        }
        emit();
    }

    private void train(Location loc) {
        if (loc.getAccuracy() > 15f) { trainStart = null; return; }
        if (trainStart == null) { startTraining(loc); return; }
        double dt = (loc.getTime() - trainStart.getTime()) / 1000.0;
        if (dt < 12) return;
        double dist = DeadReckoner.distance(trainStart.getLatitude(), trainStart.getLongitude(), loc.getLatitude(), loc.getLongitude());
        if (trainSteps >= 12 && dist >= 8 && dt <= 120) {
            double cadence = trainSteps / dt;
            double vs = Double.isNaN(trainAltStart) || Double.isNaN(baroAlt) ? 0 : (baroAlt - trainAltStart) / dt;
            // Straight-line distance undercounts on switchbacks, so only train on fairly straight walking.
            double straightness = dist / Math.max(1, trainSteps * model.predict(cadence, vs));
            if (straightness > 0.75 && model.update(cadence, vs, dist / trainSteps)) {
                prefs.edit().putString("stride", model.serialize()).apply();
            }
        }
        startTraining(loc);
    }

    private void startTraining(Location loc) {
        trainStart = loc;
        trainSteps = 0;
        trainAltStart = baroAlt;
    }

    @Override public void onStatusChanged(String p, int s, Bundle b) { }
    @Override public void onProviderEnabled(String p) { }
    @Override public void onProviderDisabled(String p) { emit(); }

    // ------------------------------------------------------------------ Sensors

    @Override public void onSensorChanged(SensorEvent e) {
        int type = e.sensor.getType();
        if (type == Sensor.TYPE_ROTATION_VECTOR || type == Sensor.TYPE_GEOMAGNETIC_ROTATION_VECTOR) {
            SensorManager.getRotationMatrixFromVector(rot, e.values);
            SensorManager.getOrientation(rot, orient);
            double h = Math.toDegrees(orient[0]);
            headingMag = (h + 360) % 360;
        } else if (type == Sensor.TYPE_PRESSURE) {
            double alt = SensorManager.getAltitude(SensorManager.PRESSURE_STANDARD_ATMOSPHERE, e.values[0]);
            long now = SystemClock.elapsedRealtime();
            if (!Double.isNaN(baroAltPrev) && now - baroPrevMs > 2000) {
                double v = (alt - baroAltPrev) / ((now - baroPrevMs) / 1000.0);
                vSpeed = 0.7 * vSpeed + 0.3 * v;
                baroAltPrev = alt; baroPrevMs = now;
            } else if (Double.isNaN(baroAltPrev)) {
                baroAltPrev = alt; baroPrevMs = now;
            }
            baroAlt = Double.isNaN(baroAlt) ? alt : 0.8 * baroAlt + 0.2 * alt;
        } else if (type == Sensor.TYPE_STEP_DETECTOR) {
            long now = SystemClock.elapsedRealtime();
            stepTimes[stepIdx % stepTimes.length] = now;
            stepIdx++;
            stepCount++;
            trainSteps++;
            if (!gpsFresh() && dr.isAnchored() && !Double.isNaN(headingMag)) {
                double cadence = cadence(now);
                double step = model.predict(cadence, vSpeed);
                dr.step(step, (headingMag + declination + 360) % 360, model.confidence(), System.currentTimeMillis());
            }
        }
    }

    private double cadence(long now) {
        int n = Math.min(stepIdx, stepTimes.length);
        if (n < 3) return 1.8;
        long oldest = stepTimes[(stepIdx - n) % stepTimes.length];
        double secs = (now - oldest) / 1000.0;
        if (secs <= 0 || secs > 20) return 1.8;
        return (n - 1) / secs;
    }

    @Override public void onAccuracyChanged(Sensor s, int a) { }

    // ------------------------------------------------------------------ Output

    private boolean gpsFresh() {
        return lastGps != null && SystemClock.elapsedRealtime() - lastGpsElapsed < GPS_STALE_MS
                && lastGps.getAccuracy() <= GOOD_GPS_ACC * 2;
    }

    private void emit() {
        try {
            JSONObject o = new JSONObject();
            long now = SystemClock.elapsedRealtime();
            boolean gpsOn = false, netOn = false;
            try { gpsOn = lm.isProviderEnabled(LocationManager.GPS_PROVIDER); } catch (Exception ignored) { }
            try { netOn = lm.isProviderEnabled(LocationManager.NETWORK_PROVIDER); } catch (Exception ignored) { }
            o.put("gpsEnabled", gpsOn);
            o.put("networkEnabled", netOn);
            o.put("heading", Double.isNaN(headingMag) ? JSONObject.NULL : (headingMag + declination + 360) % 360);
            o.put("baroAlt", Double.isNaN(baroAlt) ? JSONObject.NULL : baroAlt);
            o.put("steps", stepCount);
            o.put("strideSamples", model.samples());
            if (gpsFresh()) {
                o.put("mode", "gps");
                o.put("lat", lastGps.getLatitude());
                o.put("lon", lastGps.getLongitude());
                o.put("acc", lastGps.getAccuracy());
                o.put("alt", lastGps.hasAltitude() ? lastGps.getAltitude() : JSONObject.NULL);
                o.put("speed", lastGps.hasSpeed() ? lastGps.getSpeed() : JSONObject.NULL);
                o.put("time", lastGps.getTime());
            } else if (lastNet != null && now - lastNetElapsed < NET_STALE_MS && (!dr.isAnchored() || lastNet.getAccuracy() <= dr.radius())) {
                o.put("mode", "network");
                o.put("lat", lastNet.getLatitude());
                o.put("lon", lastNet.getLongitude());
                o.put("acc", lastNet.getAccuracy());
                o.put("time", lastNet.getTime());
            } else if (dr.isAnchored()) {
                dr.tick(System.currentTimeMillis());
                o.put("mode", "estimate");
                o.put("lat", dr.lat());
                o.put("lon", dr.lon());
                o.put("acc", dr.radius());
                o.put("walked", dr.walked());
                o.put("time", System.currentTimeMillis());
            } else {
                o.put("mode", "none");
            }
            listener.onUpdate(o);
        } catch (Exception ignored) {
        }
    }
}

File: ios/Bristlecone/Sources/LocationEngine.swift
import Foundation
import CoreLocation
import CoreMotion

/// Positioning with an honest fallback chain, as on Android:
///   1. A satellite-quality fix when one is available.
///   2. A coarse fix (Wi-Fi and cell) when it is not. Flagged as an estimate.
///   3. Dead reckoning from the last good fix using the step counter, compass and barometer,
///      with a stride length learned on this device from earlier walks. Flagged as an estimate
///      with a growing uncertainty circle.
/// iOS blends its sources itself, so the mode comes from the reported accuracy.
final class LocationEngine: NSObject, CLLocationManagerDelegate {
    private static let fixStaleMs = 12_000.0
    private static let coarseStaleMs = 60_000.0
    private static let goodAccuracy = 35.0

    let manager = CLLocationManager()
    var onUpdate: (([String: Any]) -> Void)?
    var onAuthorization: (() -> Void)?

    private let pedometer = CMPedometer()
    private let altimeter = CMAltimeter()
    private let model = StrideModel()
    private let dr = DeadReckoner()
    private var running = false
    private var timer: Timer?

    private var lastFine: CLLocation?, lastCoarse: CLLocation?
    private var lastFineAt = 0.0, lastCoarseAt = 0.0
    private var heading = Double.nan
    private var relAlt = Double.nan, absAlt = Double.nan
    private var relAltPrev = Double.nan, relAltPrevAt = 0.0
    private var vSpeed = 0.0
    private var pedSteps = 0, stepCount = 0
    private var cadenceNow = Double.nan

    // Training: steps between two clean fixes.
    private var trainStart: CLLocation?
    private var trainSteps = 0
    private var trainAltStart = Double.nan

    override init() {
        super.init()
        manager.delegate = self
        manager.desiredAccuracy = kCLLocationAccuracyBest
        manager.activityType = .fitness
        manager.headingFilter = 2
        model.deserialize(UserDefaults.standard.string(forKey: "stride"))
    }

    var hasBarometer: Bool { CMAltimeter.isRelativeAltitudeAvailable() }
    var hasStepCounter: Bool { CMPedometer.isStepCountingAvailable() }
    var strideSamples: Int { model.samples }
    func strideAt(_ cadence: Double) -> Double { model.predict(cadence: cadence, verticalSpeed: 0) }

    var authorized: Bool {
        let s = manager.authorizationStatus
        return s == .authorizedWhenInUse || s == .authorizedAlways
    }

    func resetStride() {
        model.reset()
        UserDefaults.standard.set(model.serialize(), forKey: "stride")
    }

    /// Asks for permission if needed, then starts. The answer arrives through onAuthorization.
    func request() {
        if manager.authorizationStatus == .notDetermined { manager.requestWhenInUseAuthorization() }
        else { start(); onAuthorization?() }
    }

    func start() {
        guard !running, authorized else { return }
        running = true
        manager.startUpdatingLocation()
        if CLLocationManager.headingAvailable() { manager.startUpdatingHeading() }
        if let l = manager.location, -l.timestamp.timeIntervalSinceNow < 600 {
            dr.anchor(lat: l.coordinate.latitude, lon: l.coordinate.longitude,
                      accuracy: l.horizontalAccuracy - l.timestamp.timeIntervalSinceNow, timeMs: NetCache.now())
        }
        if CMPedometer.isStepCountingAvailable() {
            pedSteps = 0
            pedometer.startUpdates(from: Date()) { [weak self] data, _ in
                guard let d = data else { return }
                DispatchQueue.main.async { self?.stepsUpdated(d) }
            }
        }
        if CMAltimeter.isRelativeAltitudeAvailable() {
            altimeter.startRelativeAltitudeUpdates(to: .main) { [weak self] d, _ in
                if let d = d { self?.altitude(d.relativeAltitude.doubleValue) }
            }
        }
        if CMAltimeter.isAbsoluteAltitudeAvailable() {
            altimeter.startAbsoluteAltitudeUpdates(to: .main) { [weak self] d, _ in
                if let d = d { self?.absAlt = d.altitude }
            }
        }
        timer = Timer.scheduledTimer(withTimeInterval: 1, repeats: true) { [weak self] _ in self?.emit() }
    }

    func stop() {
        guard running else { return }
        running = false
        manager.stopUpdatingLocation()
        manager.stopUpdatingHeading()
        pedometer.stopUpdates()
        altimeter.stopRelativeAltitudeUpdates()
        altimeter.stopAbsoluteAltitudeUpdates()
        timer?.invalidate(); timer = nil
        UserDefaults.standard.set(model.serialize(), forKey: "stride")
    }

    // MARK: CoreLocation

    func locationManagerDidChangeAuthorization(_ m: CLLocationManager) {
        if authorized { start() }
        onAuthorization?()
    }

    func locationManager(_ m: CLLocationManager, didUpdateLocations locs: [CLLocation]) {
        let now = NetCache.now()
        for l in locs where l.horizontalAccuracy >= 0 {
            if l.horizontalAccuracy <= LocationEngine.goodAccuracy * 2 {
                lastFine = l; lastFineAt = now
                if l.horizontalAccuracy <= LocationEngine.goodAccuracy {
                    dr.anchor(lat: l.coordinate.latitude, lon: l.coordinate.longitude, accuracy: l.horizontalAccuracy, timeMs: now)
                    train(l)
                }
            } else {
                lastCoarse = l; lastCoarseAt = now
                if !fineFresh && (!dr.anchored || l.horizontalAccuracy < dr.radius) {
                    dr.anchor(lat: l.coordinate.latitude, lon: l.coordinate.longitude, accuracy: l.horizontalAccuracy, timeMs: now)
                }
            }
        }
        emit()
    }

    func locationManager(_ m: CLLocationManager, didUpdateHeading h: CLHeading) {
        if h.headingAccuracy < 0 { return }
        heading = h.trueHeading >= 0 ? h.trueHeading : h.magneticHeading
    }

    func locationManager(_ m: CLLocationManager, didFailWithError error: Error) { emit() }

    // MARK: Motion

    private func stepsUpdated(_ d: CMPedometerData) {
        let total = d.numberOfSteps.intValue
        let delta = total - pedSteps
        pedSteps = total
        guard delta > 0 else { return }
        stepCount += delta
        trainSteps += delta
        if let c = d.currentCadence?.doubleValue, c > 0 { cadenceNow = c }
        if !fineFresh && dr.anchored && !heading.isNaN {
            let cadence = cadenceNow.isNaN ? 1.8 : cadenceNow
            let step = model.predict(cadence: cadence, verticalSpeed: vSpeed)
            dr.step(metres: step * Double(delta), heading: heading, confidence: model.confidence, timeMs: NetCache.now())
        }
    }

    private func altitude(_ rel: Double) {
        let now = NetCache.now()
        if relAltPrev.isNaN { relAltPrev = rel; relAltPrevAt = now }
        else if now - relAltPrevAt > 2000 {
            let v = (rel - relAltPrev) / ((now - relAltPrevAt) / 1000)
            vSpeed = 0.7 * vSpeed + 0.3 * v
            relAltPrev = rel; relAltPrevAt = now
        }
        relAlt = relAlt.isNaN ? rel : 0.8 * relAlt + 0.2 * rel
    }

    private func train(_ l: CLLocation) {
        if l.horizontalAccuracy > 15 { trainStart = nil; return }
        guard let s = trainStart else { startTraining(l); return }
        let dt = l.timestamp.timeIntervalSince(s.timestamp)
        if dt < 12 { return }
        let dist = l.distance(from: s)
        if trainSteps >= 12 && dist >= 8 && dt <= 120 {
            let cadence = Double(trainSteps) / dt
            let vs = trainAltStart.isNaN || relAlt.isNaN ? 0 : (relAlt - trainAltStart) / dt
            // Straight-line distance undercounts on switchbacks, so only train on fairly straight walking.
            let straightness = dist / max(1, Double(trainSteps) * model.predict(cadence: cadence, verticalSpeed: vs))
            if straightness > 0.75 && model.update(cadence: cadence, verticalSpeed: vs, observedStep: dist / Double(trainSteps)) {
                UserDefaults.standard.set(model.serialize(), forKey: "stride")
            }
        }
        startTraining(l)
    }

    private func startTraining(_ l: CLLocation) {
        trainStart = l
        trainSteps = 0
        trainAltStart = relAlt
    }

    // MARK: Output

    private var fineFresh: Bool { lastFine != nil && NetCache.now() - lastFineAt < LocationEngine.fixStaleMs }

    private func emit() {
        let now = NetCache.now()
        var o: [String: Any] = [
            "gpsEnabled": CLLocationManager.locationServicesEnabled() && authorized,
            "networkEnabled": CLLocationManager.locationServicesEnabled() && authorized,
            "heading": heading.isNaN ? NSNull() as Any : heading,
            "baroAlt": absAlt.isNaN ? NSNull() as Any : absAlt,
            "steps": stepCount,
            "strideSamples": model.samples
        ]
        func put(_ l: CLLocation, _ mode: String) {
            o["mode"] = mode
            o["lat"] = l.coordinate.latitude
            o["lon"] = l.coordinate.longitude
            o["acc"] = l.horizontalAccuracy
            o["time"] = l.timestamp.timeIntervalSince1970 * 1000
            if mode == "gps" {
                o["alt"] = l.verticalAccuracy >= 0 ? l.altitude as Any : NSNull()
                o["speed"] = l.speed >= 0 ? l.speed as Any : NSNull()
            }
        }
        if fineFresh, let l = lastFine {
            put(l, "gps")
        } else if let l = lastCoarse, now - lastCoarseAt < LocationEngine.coarseStaleMs, !dr.anchored || l.horizontalAccuracy <= dr.radius {
            put(l, "network")
        } else if dr.anchored {
            dr.tick(now)
            o["mode"] = "estimate"
            o["lat"] = dr.lat; o["lon"] = dr.lon; o["acc"] = dr.radius
            o["walked"] = dr.walked; o["time"] = now
        } else {
            o["mode"] = "none"
        }
        onUpdate?(o)
    }
}
